packages/design-system/src/components/primitives/dnd/use-drop.ts

```diff
@@ -0,0 +1,293 @@
+};
+
+export type UseDropHandlers = {
+  handleMove: (pointerCoordinates: { x: number; y: number }) => void;
+  handleScroll: () => void;
+  handleEnd: () => void;
+  rootRef: (target: Element | null) => void;
+};
+
+const getInitialState = <Data>() => {
+  return {
+    pointerCoordinates: undefined as { x: number; y: number } | undefined,
+    dropTarget: undefined as DropTarget<Data> | undefined,
+    childrenRectsCache: new WeakMap<UsableElement, Rect[]>(),
+    lastCandidateElement: undefined as UsableElement | undefined,
+    lastCandidateIsNearEdge: undefined as boolean | undefined,
+  };
+};
+
+export const useDrop = <Data>(props: UseDropProps<Data>): UseDropHandlers => {
+  // We want to use fresh props every time we use them,
+  // but we don't need to react to updates.
+  // So we can put them in a ref and make useMemo below very efficient.
+  const latestProps = useRef<UseDropProps<Data>>(props);
+  latestProps.current = props;
+
+  const rootRef = useRef<UsableElement>();
+  const state = useRef(getInitialState<Data>());
+
+  // We want to return a stable object to avoid re-renders when it's a dependency
+  return useMemo(() => {
+    const getChildrenRectsMemoized = (parent: UsableElement) => {
+      const fromCache = state.current.childrenRectsCache.get(parent);
+      if (fromCache !== undefined) {
+        return fromCache;
+      }
+      const result = getChildrenRects(parent);
+      state.current.childrenRectsCache.set(parent, result);
+      return result;
+    };
+
+    const setDropTarget = (partialDropTarget: PartialDropTarget<Data>) => {
+      const { pointerCoordinates } = state.current;
+      if (pointerCoordinates === undefined) {
+        return;
+      }
+
+      const parentRect = partialDropTarget.element.getBoundingClientRect();
+
+      const pointerRelativeToParent = {
+        x: pointerCoordinates.x - parentRect.left,
+        y: pointerCoordinates.y - parentRect.top,
+      };
+
+      const childrenRects = getChildrenRectsMemoized(partialDropTarget.element);
+
+      const closestChildIndex =
+        childrenRects.length === 0
+          ? 0
+          : getClosestRectIndex(childrenRects, pointerRelativeToParent);
+      const closestChildRect = childrenRects[closestChildIndex] as
+        | Rect
+        | undefined;
+
+      const childrenOrientation = getLocalChildrenOrientation(
+        partialDropTarget.element,
+        childrenRects,
+        closestChildIndex
+      );
+
+      const indexAdjustment = getIndexAdjustment(
+        pointerRelativeToParent,
+        closestChildRect,
+        childrenOrientation
+      );
+
+      const indexWithinChildren = closestChildIndex + indexAdjustment;
+
+      const current = state.current.dropTarget;
+      if (
+        current === undefined ||
+        current.element !== partialDropTarget.element ||
+        current.indexWithinChildren !== indexWithinChildren ||
+        !isEqualRect(current.rect, parentRect)
```

Comment:
this is exactly the thing I was talking about, hard to read with some fonts
```suggestion
        isEqualRect(current.rect, parentRect) === false
```